pkg/storage/segments/store.go

```diff
@@ -46,16 +46,95 @@ type Store interface {
 }
 
 type segmentStore struct {
-	oc  overlay.Client
-	ec  ecclient.Client
-	pdb pointerdb.Client
-	rs  eestream.RedundancyStrategy
+	oc            overlay.Client
+	ec            ecclient.Client
+	pdb           pointerdb.Client
+	rs            eestream.RedundancyStrategy
+	thresholdSize int
 }
 
 // NewSegmentStore creates a new instance of segmentStore
 func NewSegmentStore(oc overlay.Client, ec ecclient.Client,
-	pdb pointerdb.Client, rs eestream.RedundancyStrategy) Store {
-	return &segmentStore{oc: oc, ec: ec, pdb: pdb, rs: rs}
+	pdb pointerdb.Client, rs eestream.RedundancyStrategy, t int) Store {
+	return &segmentStore{oc: oc, ec: ec, pdb: pdb, rs: rs, thresholdSize: t}
+}
+
+// PeekThresholdReader allows a check to see if the size of a given reader
+// exceeds the maximum inline segment size or not.
+type PeekThresholdReader struct {
+	r              io.Reader
+	n              int // number of bytes read into thresholdBuf
+	thresholdBuf   []byte
+	totalReadBytes int
```

Comment:
I think we can get rid of `n` and `totalReadBytes` if we keep in `thresholdBuf` only the data that has not been read yet, and shrink the buffer whenever any data is read.

This would also greatly simplify the implementation of the Read method. See the comments below.